PetaPoco/Core/PocoData.cs

```diff
@@ -148,8 +149,16 @@ public static PocoData ForObject(object obj, string primaryKeyName, IMapper defa
                 pd.TableInfo.AutoIncrement = true;
                 foreach (var col in (obj as IDictionary<string, object>).Keys)
                 {
-                    if (col != primaryKeyName)
-                        pd.Columns.Add(col, new ExpandoColumn() { ColumnName = col });
+                    if (col == primaryKeyName) continue;
+                    if (pd.Columns.TryGetValue(col, out PocoColumn pc))
+                    {
+                        //When col and primaryKeyName only differ by case,
+                        //we need ColumnName to match case exactly.
+                        pc.ColumnName = col;
+                        continue;
+                    }
```

Comment:
What's going on here?

Reply:
Consider the following test code:
```c#
var pd = PocoData.ForType(typeof(Note), DB.DefaultMapper);
var tblNote = DB.Provider.EscapeTableName(pd.TableInfo.TableName);

CreateNote();
var entity = DB.Fetch<dynamic>($"SELECT * FROM {tblNote}").First();

DB.Update("Note", "Id", (object)entity); //<- Exception on this line
```

In the case where the custom mapper uppercase all column names, there will be a mismatch in case i.e. `entity.ID` vs the passed in primary key name `"Id"`

In PocoData,
```c#
pd.Columns = new Dictionary<string, PocoColumn>(StringComparer.OrdinalIgnoreCase);
pd.Columns.Add(primaryKeyName, new ExpandoColumn() { ColumnName = primaryKeyName });
```
we add the key as `"Id"` and also specify the `ColumnName` as `"Id"`.

So in the original for loop:
```c#
foreach (var col in (obj as IDictionary<string, object>).Keys)
{
    if (col == primaryKeyName) continue;
    pd.Columns.Add(col, new ExpandoColumn() { ColumnName = col });
}
```
`col` will never be equal to `primaryKeyName`, resulting in an `ArgumentException` with message

> An item with the same key has already been added.
since the `Dictionary` is ignoring case.

Actually I should've removed this line. It's been replaced with the key lookup:
```c#
if (col == primaryKeyName) continue;
```

So this change fixes the duplicate key bug by making sure the `ColumnName` conforms to the mapper logic.
For consistency we should technically also update the `Key`, but it's immutable and the `Dictionary` is ignoring case anyway.

Reply:
Hmm, I need to think on this a little more. I get what you're saying, but something about it doesn't seem right to me. Maybe the primary key name should be run through the mapper? Maybe other fields need to be run through the mapper? Maybe it's just the caller's responsibility to pass in the correctly cased primary key name? After all, this param is not supposed to be the name of the field on the POCO; it's supposed to literally be the name of the table field.

I use an underscore mapper in my classes, so that a POCO property called `PersonID` gets mapped to a db field called `person_id`. If I'm calling that `Update()` overload, I know that I need to specify the primary key name as `'person_id`, not `PersonID`.

Reply:
To be honest, I got confused most of the time about when the user should use poco casing or db casing, while working on this. So you're probably right.

My initial thought was that the dictionary keys represent the poco case (even though `PropertyInfo` is there) to help dynamic/anonymous marry strongly typed pocos, since this case won't have `PropertyInfo`, and `ColumnName` represents db case. Since that approach didn't cover all the cases, the dictionary was made to be case insensistive.

Since this part of the code is purely for dynamic, and the only affected property is the primary key, we can go both ways. Either escape the primary key (mapper should ensure original code hits exact match) or leave it to the user as long as we state it somewhere in the documentation (which I'm sure it's there somewhere)

How would you like to proceed?

Reply:
While overriding the tests I saw this line of code, which should throw if I apply the logic we discussed regarding the primary key i.e. user must provide correct casing:
```c#
DB.UpdateAsync("Note", "ID", entity); //Note: TableName is not capitalized
```

So it seems we force the user to use correct casing for primary key in order to get `PocoData`, but we escape both table name and primary key (when building the sql statement) [here](https://github.com/CollaboratingPlatypus/PetaPoco/blob/53e928c6dc44819aa834683abcd108029f960b77/PetaPoco/Database.cs#L2506C12-L2507):
```c#
cmd.CommandText =
    $"UPDATE {_provider.EscapeTableName(tableName)} SET {sb} WHERE {_provider.EscapeSqlIdentifier(primaryKeyName)} = {_paramPrefix}{index++}";
```


> Maybe the primary key name should be run through the mapper? Maybe other fields need to be run through the mapper?

Now it looks more like we should run primary key AND other fields through the mapper.

Reply:
I need to put together a couple of small concrete tests to step through this, but in principle I disagree. Mappers mediate between POCO class/property names and database table/column names. On the other hand, generally speaking, it is the caller's responsibility to make sure that anything that goes into a PetaPoco function as a string will be understood by the database. (Mappers don't peek into SQL strings, for example.) A `dynamic` object doesn't have fixed properties to map to, and so the mappers have no role.

Remember, too, that while the mapping from property name to column name is deterministic, the mapping from column name to property name is not. That is, a mapper which uppercases will always map a property called `Id` into a column called `ID`. But given a column called `ID`, there is no way to know whether the POCO property was `Id`, `id`, or `ID`. 

So I think the existing code in `PocoData.ForObject()` is correct. The incoming `obj`, as an Expando, will have received the column names as they are in the database, and in a call to something like `DB.Update()` the user is expected to pass in properly cased table and primary key names.

Reply:
Agreed. It doesn't make sense to escape `primaryKeyName` since it's not deterministic. We also assume `obj` to contain keys matching the database. There's also no guarantee that it contains the primary key.

Base on the above and the fact that the dictionary is case insensitive, I see no harm in using `TryGetValue` or `ContainsKey` to avoid adding duplicates, although the former also gives you the opportunity to fix the casing of the primary key column based on `obj`.

Are you concerned about performance?

Reply:
I've undone the PocoData changes as requested, will make them again in another branch based on develop.
Do we resolve this and continue the discussion on the coming PR?